force-app/main/default/classes/IBMWatsonIAMTokenManager.cls

```diff
@@ -0,0 +1,167 @@
+      // refresh current token
+      token = refreshToken();
+    } else {
+      // use valid managed token
+      token = tokenData.getAccessToken();
+    }
+
+    return token;
+  }
+
+  /**
+   * Request an IAM token using an API key. Also updates internal managed IAM token information.
+   *
+   * @return the new access token
+   */
+  private String requestToken() {
+    IBMWatsonRequestBuilder builder = IBMWatsonRequestBuilder.httpPost(url);
+
+    builder.addHeader(IBMWatsonHttpHeaders.CONTENT_TYPE, IBMWatsonHttpMediaType.APPLICATION_FORM_URLENCODED);
+    builder.addHeader(IBMWatsonHttpHeaders.AUTHORIZATION, DEFAULT_AUTHORIZATION);
+
+    IBMWatsonFormBody formBody = new IBMWatsonFormBody.Builder()
+      .add(GRANT_TYPE, REQUEST_GRANT_TYPE)
+      .add(API_KEY, apiKey)
+      .add(RESPONSE_TYPE, CLOUD_IAM)
+      .build();
+    builder.body(formBody);
+
+    tokenData = callIamApi(builder.build());
+    return tokenData.getAccessToken();
+  }
+
+  /**
+   * Refresh an IAM token using a refresh token. Also updates internal managed IAM token information.
+   *
+   * @return the new access token
+   */
+  private String refreshToken() {
+    IBMWatsonRequestBuilder builder = IBMWatsonRequestBuilder.httpPost(url);
+
+    builder.addHeader(IBMWatsonHttpHeaders.CONTENT_TYPE, IBMWatsonHttpMediaType.APPLICATION_FORM_URLENCODED);
+    builder.addHeader(IBMWatsonHttpHeaders.AUTHORIZATION, DEFAULT_AUTHORIZATION);
+
+    IBMWatsonFormBody formBody = new IBMWatsonFormBody.Builder()
+      .add(GRANT_TYPE, REFRESH_GRANT_TYPE)
+      .add(REFRESH_TOKEN, tokenData.getRefreshToken())
+      .build();
+    builder.body(formBody);
+
+    tokenData = callIamApi(builder.build());
+    return tokenData.getAccessToken();
+  }
+
+  /**
+   * Check if currently stored access token is expired.
+   *
+   * Using a buffer to prevent the edge case of the
+   * token expiring before the request could be made.
+   *
+   * The buffer will be a fraction of the total TTL. Using 80%.
+   *
+   * @return whether the current managed access token is expired or not
+   */
+  private boolean isAccessTokenExpired() {
+    if (tokenData.getExpiresIn() == null || tokenData.getExpiration() == null) {
+      return true;
+    }
+
+    Double fractionOfTimeToLive = 0.8;
+    Long timeToLive = tokenData.getExpiresIn();
+    Long expirationTime = tokenData.getExpiration();
+    Double refreshTime = expirationTime - (timeToLive * (1.0 - fractionOfTimeToLive));
+    Double currentTime = Math.floor(System.now().getTime() / 1000);
+
+    return refreshTime < currentTime;
+  }
+
+  /**
+   * Used as a fail-safe to prevent the condition of a refresh token expiring,
+   * which could happen after around 30 days. This function will return true
+   * if it has been at least 7 days and 1 hour since the last token was
+   * retrieved.
+   *
+   * @returns whether the current managed refresh token is expired or not
+   */
+  private boolean isRefreshTokenExpired() {
```

Comment:
not quite sure I understand the logic here but if it is a best practice or advised by others that's fine. 7 days seems somewhat arbitrary. 

Also I'm slightly unclear on how this refresh token issue would ever be invoked unless the refresh token is being managed via the user of the SDK in a separate storage (which doesn't seem to be the case since the refresh token has `private` visibility)

but no issues here for me

Reply:
The 7 days is pretty arbitrary, but after it was chosen it was used across the other SDKs, so we might as well be consistent.

As for the refresh token issue itself, we added the method because we use the refresh token to get a new access token when it expires. What we were worried about was trying to use the refresh token when that has _also_ expired. That happens after 30 days, but just to try and proactively make sure we don't have that problem, we'll get a new token after 7 days.